src/Filter/QueryParameterValidateListener.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Filter;
+
+use ApiPlatform\Core\Api\FilterLocatorTrait;
+use ApiPlatform\Core\Exception\FilterValidationException;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Validates query parameters depending on filter description.
+ *
+ * @author Julien Deniau <julien.deniau@gmail.com>
+ */
+class QueryParameterValidateListener
+{
+    use FilterLocatorTrait;
+
+    private $resourceMetadataFactory;
+
+    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, ContainerInterface $filterLocator)
+    {
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+        $this->setFilterLocator($filterLocator);
+    }
+
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        if (
+            !$request->isMethodSafe(false)
+            || !($attributes = RequestAttributesExtractor::extractAttributes($request))
+            || !isset($attributes['collection_operation_name'])
+            || 'get' !== ($operationName = $attributes['collection_operation_name'])
+        ) {
+            return;
+        }
+
+        $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);
+        $resourceFilters = $resourceMetadata->getCollectionOperationAttribute($operationName, 'filters', [], true);
+
+        foreach ($resourceFilters as $filterId) {
+            if (!$filter = $this->getFilter($filterId)) {
+                continue;
+            }
+
+            foreach ($filter->getDescription($attributes['resource_class']) as $name => $data) {
+                $errorList = [];
+
+                if (($data['required'] ?? false) && null === $request->query->get($name)) {
```

Comment:
This condition will not work with serializer's filters (assuming they are required) for example. Indeed, `property[]` is not equal to `property`.

Reply:
Sorry but I did not really understand what you mean here ?

Can you provide an example when this does not work ?

Reply:
Let's take the description of the property filter: https://github.com/api-platform/core/blob/63d555ae0020dc6c66feef86ff56ac5131b1f7b3/src/Serializer/Filter/PropertyFilter.php#L68-L77
And let's say `required` is equal to `true` for our example.

So, the `$name` var will be equal to `property[]` here. But with `?property['relatedDummy'][]=name` the name will be `property` in the parameter bag and `$request->query->get('property[]')` will return `null`...